book/src/proposals/durable-tx-nonces.md

```diff
@@ -23,30 +23,59 @@ implemented by some CPU ISAs.
 
 ### Contract Mechanics
 
-TODO: svgbob this into a flowchart, add prose
+TODO: svgbob this into a flowchart
 
 ```text
 Start
 Create Account
   state = Uninitialized
-InitializeInstruction(hash)
-  if hash is recent
-    stored_hash = hash
+InitializeInstruction()
+  if !sysvar.recent_blockhashes.is_empty()
+    stored_hash = sysvar.recent_blockhashes[0]
     state = Initialized
     success
   else
     error
-NonceInstruction(spend_hash, next_hash)
+NonceInstruction(spend_hash) /* spend_hash is a stand in for a TBD sysvar
+                                which will replace this "parameter" */
   if spend_hash == stored_hash
-    if next_hash is recent
-      stored_hash = next_hash
+    if !sysvar.recent_blockhashes.is_empty()
+      stored_hash = sysvar.recent_blockhashes[0]
       success
     else
       error
   else
     error
 ```
 
+A client wishing to use this feature starts by creating a system account and
+depositing sufficient lamports as to make it rent-exempt. The resultant account
+will be in the `Uninitialized` state with no stored hash and thus unusable.
+
+To begin using the account an `Initialize` instruction is executed on it,
+advancing its state to `Initialized` and storing a durable nonce, chosen by the
+contract from the `recent_blockhashes` sysvar, in the data field.
+
+When making use of a durable nonce, the client must first query its value from
+account data. A transaction is now constructed in the normal way, but with the
+following additional requirements:
+
+ 1) The durable nonce value is used in the `recent_blockhash` field
+ 2) A `Nonce` instruction is issued (first?)
```

Comment:
last.  the runtime must verify that the Nonce contract holds a matching recent_blockhash

Reply:
Sorry, this is a little ambiguous.  I meant "first" in relation to other instructions in the TX.  Gas savings and whatnot

Reply:
ok, yeah, the runtime can look at the Nonce account before issuing the first instruction

can we require that the Nonce() instruction be first?  seems awkward.

Reply:
> can we require that the Nonce() instruction be first?

That's what I'm getting at. Yeah.

> seems awkward.

Awkward how? Seems like knowing exactly where to check for the nonce instruction and enforcing that nonce instruction has an opportunity to fail first are both GoodThing(TM)s

Reply:
ok, I think you're saying that this will be what the runtime does (from convo above)

Are we planning to collect the TX fee even if the blockhash doesn't match the Nonce?  Sounds bad (replay TX until payer is out of tokens...)

Reply:
Hrm, yeah I think you're right about the fees.  This happens when *any* instruction fails in a transaction 🤔.  It seems like *not* charging the fee could lead to DoS of the validators though.

Reply:
What if `Nonce` does the full withdraw, making the account single-use?

Reply:
not enforceable, nor really possible under current runtime rules, and the runtime needs to defend against spam